### ISSUE TITLE
LADSPA Documentation

### ISSUE THREAD

**Issue by LordKobra** [sadko4u/lsp-plugins#284]
Hello Vlad,

thanks a lot for this great plugin tool! It's probably just what i need for a system wide parametric equalizer. I tried to load "Parametric Equalizer x16 Stereo" in pulseaudio using `pacmd load-module module-ladspa-sink`, but the "control" part is giving me some issues. Is there any documentation on how to set these parameters? For example when i export the GUI parameters into a .cfg file, the names are different and there seem to exist less parameters, than when i compare it to the LADSPA version with `analyseplugin` (e.g. "Input signal meter Left (G)" does only exist in ladspa). I am quite new to this and sadly don't understand a lot, so is there a way to parse my .cfg values for use with the ladspa version?

Best regards!

**Comment by sadko4u** [comment]
Hello!

The parameters can be split into two basic part: input parameters for configuration and output parameters, like meters, for monitoring.
When LSP plugin saves the configuration file, it saves the whole set of input parameters, so you don't need to worry about that *.cfg file is missing something.
You also can use the standalone JACK application instead of LADSPA to configure the plugin properly, then export the configuration and apply it in your setup.

For the gain values, you should consider that gain values are just multipliers, not 'dB' values. That means, that gain of value `0` is `-inf dB` and gain of value `1` is `0 dB`. But since 1.2.0 release configuration files are more user-friendly and you can write something like this in your configuration file:
```
g_out = -6 db
```

If you have any additional questions, you're welcome.

**Comment by sadko4u** [comment]
Demonstrations of usage of all of the plugins are available here:
https://youtube.com/playlist?list=PL_kZt3RHXWbmWCNDrdjrVM--kiENuhkKV

**Comment by LordKobra** [comment]
Thank you for the answers! I realize my first message was a bit messy, so i will try and clarify my issues.

The general goal is to use a parametric equalizer to test my headphone according to a configuration file in the bottom left table of this link: https://www.dropbox.com/s/l0ucv96zjjp8cvv/AKG%20K712.pdf?dl=0

For some dependency reasons on my distro, i am limited to pulseaudio. From what i understood, both JACK and Pipewire serve as replacements for pulseaudio, so i can't use them and have to resort to loading the plugins in pulseaudio.

Currently my approach looked like this: I set the values in the LSP GUI, export it to a .cfg file and manually convert the values to be conform for the following example command:
```
pacmd load-module module-ladspa-sink sink_name=lsp_k712 sink_master=myadaptername plugin=lsp-plugins-ladspa-1.2.3 label="http://lsp-plug.in/plugins/ladspa/para_equalizer_x16_stereo" control=0,0.75,1.0,0,0,0.2,0.0,0.015851,1,0.0,0.0,1,1,1,0,0,0,0,28.0,3.0,1.2,0,5,0,0,0,0,105.0,5.5,0.71,0.0625,3,0,0,0,0,690.0,7.0,0.71,0.125,0,0,0,0,0,1900.0,0.1,3.0,0.18750,0,0,0,0,0,2250.0,0.1,2.5,0.25,1,0,0,0,0,2900.0,4.5,1.0,0.3125,0,0,0,0,0,3900.0,0.1,4.0,0.375,0,0,0,0,0,5700.0,0.1,1.5,0.4375,1,0,0,0,0,7600.0,0.1,5.0,0.5,0,0,0,0,0,11000.0,0.1,0.71,0.5625,0,0,0,0,0,1600.0,0.1,0.0,0.625,0,0,0,0,0,2500.0,0.1,0.0,0.6875,0.1,0,0,0,0,4000.0,0.1,0.0,0.75,0,0,0,0,0,6300.0,0.1,0.0,0.8125,0,0,0,0,0,10000.0,0.1,0.0,0.875,0,0,0,0,0,16000.0,0.1,0.0,0.9375
```

It's not really working without issues, so i want to ask a few questions:

1.) Can i directly read the config file with the hooked plugin, so does it seek a .cfg file at a specific path internally? Then i could just copy the exported .cfg file and rename it. (Annotation: If i skip the `control=...` section, the command will return with an error right now.)

2.) If not, how can i convert `gain=-8.5db` to the multiplicative scale? Does there exist a correct formula?

3.) In the PDF, a filter called "PEAK" is mentioned, probably for EQ APO. Does this translate to "Bell" in lsp-plugins?

Best regards!

**Comment by sadko4u** [comment]
1. No, this is not possible since the host contols the values for input parameters of LADSPA plugin. LADSPA plugin can not be an initiator to adjust it's parameters. I don't see any another hack except of writing another one wrapper around the LADSPA plugin that will load the desired plugin and set up the desired input port values of the wrapped plugin.

2. Yes, that's simple. Here's are the code snippets:

https://github.com/lsp-plugins/lsp-dsp-units/blob/d312ab708c2b9c6e2a51ccca1c5ddfed4e24bff0/include/lsp-plug.in/dsp-units/units.h#L117-L125

https://github.com/lsp-plugins/lsp-dsp-units/blob/d312ab708c2b9c6e2a51ccca1c5ddfed4e24bff0/include/lsp-plug.in/dsp-units/units.h#L137-L145

4. The APO `PEAK` is tranlated into `BELL` with the `APO DR` filter mode enabled.

https://github.com/lsp-plugins/lsp-plugins-para-equalizer/blob/49901217a5427c39ac322d0678129a68d08e6b3a/src/main/ui/para_equalizer.cpp#L295-L301

**Comment by LordKobra** [comment]
Okay, that's already good to hear!

1.) It's not such a big issue, i could try to write a little format conversion cfg->control. However i've found and tested another parametric EQ, called DSP, and it works as LADSPA plugin reading from a config file. I don't know if this is done with a wrapper, but it might be worth taking a look: https://github.com/bmc0/dsp/wiki/System-Wide-DSP-Guide

2., 3.) Thanks a lot, i will try these conversions as soon as i got time and report back if everything works.

**Comment by LordKobra** [comment]
I got the conversion to work writing a small python script! I tested it against dsp, and there it sounds slightly different. Maybe it's just a difference in filters (as dsp doesn't guarantee APO EQ equivalents anywhere), or i still got a bug in my script. On the weekend, I'll manually check what i did wrong and also check how APO EQ sounds on a Windows machine.

Until then, one last question came to my mind:
1) Can the visual interface in the GUI version display the filter curve without using JACK? I just need the EQ Curve, like in the table "EQ Curve - total" https://www.dropbox.com/s/l0ucv96zjjp8cvv/AKG%20K712.pdf?dl=0 . Right now the spectrum graph is empty.

Best regards!

**Comment by sadko4u** [comment]
No, the plugin design relies on the transport from the DSP part to the UI part which is performed in the run() callback method. But you can use JACK dummy driver if you don't want to acquire your audio device by the JACK server. Plugins work with the dummy driver pretty well.

**Comment by sadko4u** [comment]
Hello! Nice to see your problem has been solved.
